src/security/builtin_plugins/cryptographic/src/crypto_key_factory.c

```diff
@@ -0,0 +1,1872 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <assert.h>
+#include <string.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
+#include <openssl/hmac.h>
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include "dds/ddsrt/atomics.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/sync.h"
+#include "dds/ddsrt/types.h"
+#include "dds/security/dds_security_api.h"
+#include "dds/security/core/dds_security_utils.h"
+#include "dds/security/core/dds_security_serialize.h"
+#include "dds/security/core/shared_secret.h"
+#include "crypto_defs.h"
+#include "crypto_utils.h"
+#include "crypto_cipher.h"
+#include "crypto_key_factory.h"
+#include "crypto_objects.h"
+
+/**
+ * Implementation structure for storing encapsulated members of the instance
+ * while giving only the interface definition to user
+ */
+
+#define KXKEYCOOKIE "key exchange key"
+#define KXSALTCOOKIE "keyexchange salt"
+
+typedef struct dds_security_crypto_key_factory_impl
+{
+  dds_security_crypto_key_factory base;
+  const dds_security_cryptography *crypto;
+  ddsrt_mutex_t lock;
+  struct CryptoObjectTable *crypto_objects; /* table for storing CryptoHandle - ParticipantKeyMaterial pairs */
+  ddsrt_atomic_uint32_t next_key_id;
+} dds_security_crypto_key_factory_impl;
+
+static void
+crypto_token_copy(
+    master_key_material *dst,
+    const DDS_Security_KeyMaterial_AES_GCM_GMAC *src)
+{
+  DDS_Security_CryptoTransformKind_Enum src_transform_kind = CRYPTO_TRANSFORM_KIND(src->transformation_kind);
+
+  if (CRYPTO_TRANSFORM_HAS_KEYS(dst->transformation_kind) && !CRYPTO_TRANSFORM_HAS_KEYS(src_transform_kind))
+  {
+    ddsrt_free(dst->master_salt);
+    ddsrt_free(dst->master_sender_key);
+    ddsrt_free(dst->master_receiver_specific_key);
+  }
+  else if (CRYPTO_TRANSFORM_HAS_KEYS(src_transform_kind))
+  {
+    uint32_t key_bytes = CRYPTO_KEY_SIZE_BYTES(src_transform_kind);
+    if (!CRYPTO_TRANSFORM_HAS_KEYS(dst->transformation_kind))
```

Comment:
Is it a given that ``src`` and ``dst`` have the same key size if both have keys?

Reply:
No, it's not, so I've changed the code so that it always frees the memory and allocates new buffers (which has minor performance impact because `crypto_token_copy` is only called during handshake/initialization)